docs/src/userguide/numpy_tutorial.rst

```diff
@@ -187,62 +189,95 @@ compatibility. Here's :file:`convolve_typed.pyx`. *Read the comments!*
-of C code to set up while in :file:`convolve2.c` a normal C for loop is used.
+Especially have a look at the for loops: In :file:`convolve_cy.c`, these are ~20 lines
+of C code to set up while in :file:`convolve_typed.c` a normal C for loop is used.
 
 After building this and continuing my (very informal) benchmarks, I get:
 
 .. sourcecode:: ipython
 
-    In [21]: import convolve2
-    In [22]: %timeit -n2 -r3 convolve2.naive_convolve(f, g)
-    2 loops, best of 3: 828 ms per loop
+    In [19]: %timeit -n2 -r3 convolve_py.naive_convolve(f, g)
+    3.9 s ± 12.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+    In [20]: %timeit -n2 -r3 convolve_cy.naive_convolve(f, g)
+    3.12 s ± 15.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
+    In [22]: %timeit -n2 -r3 convolve_typed.naive_convolve(f, g)
+    13.8 s ± 122 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
 
-Efficient indexing
-====================
+So in the end, adding types make the Cython code slower?
+
+What happend is that most of the time spend in this code is spent on line
+60. ::
+
+    value += g[smid - s, tmid - t] * f[v, w]
+
+So what made this line so much slower than in the pure Python version?
+
+``g`` and ``f`` are still NumPy arrays, so Python objects, and expect
+Python integers as indexes. Here we give C integers. So every time
+Cython reaches this line, it has to convert all the C integers to Python
+integers. Since this line is called very often, it outweight the speed
+benefits of the pure C loops that were created from the ``range()`` earlier.
+
+Furthermore, ``g[smid - s, tmid - t] * f[v, w]`` returns a Python integer
+and ``value`` is a C integers, Cython has to do a type conversion again.
+In the end those types conversions add up. And made our convolution really
+slow. But this can be solved easily by using memoryviews.
+
+Efficient indexing with memoryviews
+===================================
 
-There's still a bottleneck killing performance, and that is the array lookups
-and assignments. The ``[]``-operator still uses full Python operations --
+There are still two bottleneck killing performance, and that is the array lookups
+and assignments, as well as C/Python types conversion.
+The ``[]``-operator still uses full Python operations --
 what we would like to do instead is to access the data buffer directly at C
 speed.
 
 What we need to do then is to type the contents of the :obj:`ndarray` objects.
-We do this with a special "buffer" syntax which must be told the datatype
-(first argument) and number of dimensions ("ndim" keyword-only argument, if
-not provided then one-dimensional is assumed).
+We do this with a memoryview. There is :ref:`a page in the Cython documentation
+<memoryviews>` dedicated to it.
 
-More information on this syntax [:enhancements/buffer:can be found here].
+In short, memoryviews are C structures that can hold a pointer to the data
+of a NumPy array. They also support slices, so they work even if
```

Comment:
Well, a pointer, yes, but also all the necessary buffer metadata for provide efficient access: dimensions, strides, item size, item type information, etc.